python_modules/dagster/dagster/core/execution/context/compute.py

```diff
@@ -237,6 +247,48 @@ def get_tag(self, key: str) -> Optional[str]:
         """
         return self._step_execution_context.get_tag(key)
 
+    def has_events(self) -> bool:
+        return bool(self._events)
+
+    def retrieve_events(self) -> Iterator[DagsterEvent]:
+        while self._events:
+            yield self._events.pop(0)
+
+    def log_event(self, event) -> None:
+        if isinstance(event, (AssetMaterialization, Materialization)):
+            self._events.append(
+                DagsterEvent.asset_materialization(
+                    self._step_execution_context,
+                    event,
+                    self._step_execution_context.get_input_lineage(),
+                )
+            )
+        elif isinstance(event, AssetObservation):
+            self._events.append(DagsterEvent.asset_observation(self._step_execution_context, event))
+        elif isinstance(event, ExpectationResult):
+            self._events.append(
+                DagsterEvent.step_expectation_result(self._step_execution_context, event)
+            )
+        else:
+            check.failed("Unexpected event {event}".format(event=event))
+
+    def log_metadata_for_output(self, metadata: Dict[str, Any], output_name: Optional[str] = None):
```

Comment:
@alangenfeld - any thoughts on the naming here? Is "log" appropriate? Would `log_output_metadata`, which is slightly less verbose, be less clear?

Also, @dpeng817 - can we use use `Mapping` instead of `Dict`? Might require changing the parameter annotation for the `Output` constructor.